### DISCUSSION TITLE
Pass additional context to the azure pipelines manager to determine the current Azure DevOps project

### DISCUSSION THREAD

**Discussion by robjackstewart** [renovatebot/renovate#27017] [Suggest an Idea]
When an azure pipelines declares a Azure Repos repository resource, the name of the repository may be prefixed with the Azure DevOps project name e.g. `repo: Project/Repository`. However, the project name is not mandatory and omitting it makes the reference to the current project implicit.

For example:
If a repository called `MyRepository` is in Azure DevOps project `MyProject`, and there is an Azure Pipelines YAML file which declares a repository resource as `repo: MyOtherRepository`, then Azure Pipelines assumes that `MyOtherRepository` is in project `MyProject`. Whereas if `MyOtherRepository` was in a different project it would need to be declared as `repo: MyOtherProject/MyOtherRepository`.

Currently, renovate does cannot update dependencies if the project is implicit because we do not pass in context about the repository being processed to the manager. I would like pass the name of the current repository being 

To introduce this case in a generic manner that may be adapted for other managers in future, I propose that we introduce manager specific config interfaces that inherit from `ExtractConfig` which contain any additional context or information required for the manager to function. We build this instance based on the name of the manager currently being processed and extract information from the `managerConfig`, which itself has access to the renovate config information, and then pass this new instance to the `extractAllPackageFiles` and `extractAllPackageFile` methods in as the config parameter.

This requires passing the `managerConfig` in addition to the `workerExtractConfig` to the `getManagerPackageFiles` function such that the manager specific extract config can be built prior to calling the `extractAllPackageFiles` and `extractAllPackageFile` methods.

I am not suggesting that we introduce manager specific extract config by default, but only if and when the manager needs additional information.

I have made and tested these changes, at least in the case of azure pipelines, [here](https://github.com/renovatebot/renovate/compare/main...robjackstewart:renovate:add-azure-pipeline-project-calc). I've also changed the strings used to register the managers into an `enum` such that they may be referenced in the `buildManagerSpecificExtractConfig`. Naturally, my changes only account for the `azure-pipelines` manager in the `buildManagerSpecificExtractConfig` function but I envisage this can become a `switch` statement if any other managers need access to additional information available in the `managerConfig` or `workerExtractConfig` with the respective `case` using the `ManagerKeys` enum.

May I please get some feedback on this proposal and [the sample changes I have made so far](https://github.com/renovatebot/renovate/compare/main...robjackstewart:renovate:add-azure-pipeline-project-calc)? 😄

**Comment by robjackstewart** [comment]
There was some discussion on the implementation of this in #15028. Linking for relevance.

**Comment by rarkins** [comment]
We prefer to avoid putting manager-specific logic into lib/workers code. Did you consider adding `repository` to `WorkerExtractConfig`?

**Reply by robjackstewart** [reply]
Thats cool, totally understand. It would need to be added to `ExtractConfig` because that's the interface that is currently expected by the managers. Even though its the `WorkerExtractConfig` that gets passed under the hood. I'll update my changes and get back to you.

**Reply by robjackstewart** [reply]
@rarkins Ive updated my implementation to add `repository?: string` to the `ExtractConfig` interface. [Does this look more acceptable](https://github.com/renovatebot/renovate/compare/main...robjackstewart:renovate:add-azure-pipeline-project-calc)?

**Reply by viceice** [reply]
oh, you need the current org/repo string? 🤔 it seems the project is the equivalent of GitHub organization? or more like a gitlab sub group? 🤔

@rarkins @secustor maybe we can have some global config for repo stage? so we could do `RepoConfig.get` like `GlobalConfig.get` where we can get repo url and other similar Infos returned from init repo. so that can later be reused by managers when working on same platform/ host / org / repo...

**Reply by rarkins** [reply]
Any downside to passing it in extract config?

**Comment by viceice** [comment]
I think the manager can simply use the platform url from global config. so no need to pass it down.

**Reply by robjackstewart** [reply]
Thanks for the suggestion! I can't see any keys in the `GlobalConfig` that suggest I have access to the platform URL. the `'platform'` key will just return a value of  `'azure'` and the `'endpoint'` will return a value like `https://dev.azure.com/<organisation-name>/` which is missing the project. May you please elaborate on your suggestion? 😸

**Comment by robjackstewart** [comment]
[PR raised](https://github.com/renovatebot/renovate/pull/27277)
